### PR TITLE
Add recipes for MongoDb related changes in Spring Boot/Data 2.3

### PR DESCRIPTION
**Pull request by Laurens-W**
## What's changed?

Added a few refactoring steps related to mongoDb including a testcase

## Anyone you would like to review specifically?

@timtebeek 

## Any additional context

https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-2.3-Release-Notes#spring-data-neumann
https://spring.io/blog/2020/05/18/what-s-new-in-spring-data-klara-dan-von-neumann
https://docs.spring.io/spring-data/mongodb/docs/3.0.0.RELEASE/reference/html/#upgrading.2-3

### Checklist
- [x] I've added unit tests to cover both positive and negative cases
- [x] I've read and applied the [recipe conventions and best practices](https://docs.openrewrite.org/authoring-recipes/recipe-conventions-and-best-practices)
- [x] I've used the IntelliJ IDEA auto-formatter on affected files

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -121,8 +121,11 @@ dependencies {
     implementation("org.openrewrite:rewrite-yaml")
     implementation("org.openrewrite:rewrite-gradle")
     implementation("org.openrewrite:rewrite-maven")
+    implementation("org.openrewrite:rewrite-templating:$rewriteVersion")
     implementation("org.openrewrite.recipe:rewrite-java-dependencies:${rewriteVersion}")
     implementation("org.openrewrite.gradle.tooling:model:${rewriteVersion}")
+    implementation("org.springframework.data:spring-data-mongodb:2.2.13.RELEASE")
+    implementation("org.mongodb:mongo-java-driver:3.12.14")
```
Comment (timtebeek) [inline build.gradle.kts:128]: I think these can be `compileOnly`, as the recipes we generate will only contain references in Strings; might need to be checked against a project in practice.

File: src/main/resources/META-INF/rewrite/spring-data-23.yml [hunk 2]
```diff
@@ -23,3 +23,18 @@ displayName: Migrate to Spring Data 2.3
 description: Migrate applications to the latest Spring Data 2.3 release.
 recipeList:
   - org.openrewrite.java.spring.data.MigrateJpaSort
+  - org.openrewrite.java.dependencies.AddDependency:
+      groupId: org.mongodb
+      artifactId: mongodb-driver-legacy
+      version: "5.1.x"
```
Comment (timtebeek) [inline src/main/resources/META-INF/rewrite/spring-data-23.yml:29]:
```suggestion
      version: 5.1.x
```

File: src/testWithSpringBoot_2_3/java/org/openrewrite/java/spring/data/RefactorSimpleMongoDbFactoryTest.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+              """,
+            """
+              import org.springframework.data.mongodb.MongoDatabaseFactory;
+              import org.springframework.data.mongodb.core.SimpleMongoClientDatabaseFactory;
+
+              class Test {
+                  private MongoDatabaseFactory factory;
+                  private SimpleMongoClientDatabaseFactory factory2;
+
+                  public void setupUri(String uri) {
```
Comment (timtebeek) [inline src/testWithSpringBoot_2_3/java/org/openrewrite/java/spring/data/RefactorSimpleMongoDbFactoryTest.java:62]:
```suggestion
                  void setupUri(String uri) {
```

File: src/testWithSpringBoot_2_3/java/org/openrewrite/java/spring/data/RefactorSimpleMongoDbFactoryTest.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+                  }
+              }
+              """,
+            """
+              import org.springframework.data.mongodb.MongoDatabaseFactory;
+              import org.springframework.data.mongodb.core.SimpleMongoClientDatabaseFactory;
+
+              class Test {
+                  private MongoDatabaseFactory factory;
+                  private SimpleMongoClientDatabaseFactory factory2;
```
Comment (timtebeek) [inline src/testWithSpringBoot_2_3/java/org/openrewrite/java/spring/data/RefactorSimpleMongoDbFactoryTest.java:60]:
```suggestion
                  MongoDatabaseFactory factory;
                  SimpleMongoClientDatabaseFactory factory2;
```

File: build.gradle.kts (hunk 1 above)
Comment (timtebeek) [reply build.gradle.kts]: Did you check this against a small real project? Just to be sure that part was not missed.

**Comment by timtebeek** [comment]
Also: could you add a link to the relevant migration guide that made you place this in Spring Boot 2.3 ?
I think that might be helpful to have here, both in review, and if it comes up in the future.

**Comment by okundzich** [comment]
If we are unsure where to place these recipes, I think just create them and provide them so they can bundle into their own recipe.

**Comment by Laurens-W** [comment]
@timtebeek @okundzich I've added some relevant links as additional context. Spring Boot 2.3 came with the so called `spring data neumann` release which upgraded spring-data-mongodb from 2.2 to 3. The reference docs for the 3.0.0 release mention these changes (and more) as deprecations/api removals. The only part I'm unsure about is adding the legacy driver with version 5.1.

File: build.gradle.kts (hunk 1 above)
Comment (Laurens-W) [reply build.gradle.kts]: Will do!

File: src/main/resources/META-INF/rewrite/spring-data-23.yml [hunk 5]
```diff
@@ -23,3 +23,18 @@ displayName: Migrate to Spring Data 2.3
...
+  - org.openrewrite.java.spring.data.RefactorSimpleMongoDbFactoryRecipe
+  - org.openrewrite.java.ChangeType:
+      oldFullyQualifiedTypeName: org.springframework.data.mongodb.MongoDbFactory
+      newFullyQualifiedTypeName: org.springframework.data.mongodb.MongoDatabaseFactory
+  - org.openrewrite.java.ChangeType:
+      oldFullyQualifiedTypeName: org.springframework.data.mongodb.core.SimpleMongoDbFactory
+      newFullyQualifiedTypeName: org.springframework.data.mongodb.core.SimpleMongoClientDatabaseFactory
+  - org.openrewrite.java.ChangeType:
+      oldFullyQualifiedTypeName: org.springframework.data.mongodb.core.SimpleMongoClientDbFactory
+      newFullyQualifiedTypeName: org.springframework.data.mongodb.core.SimpleMongoClientDatabaseFactory
```
[the remark this answers is not in the document]
Comment (Laurens-W) [reply src/main/resources/META-INF/rewrite/spring-data-23.yml:40]: The `Options` to `Settings` factory seems a little more hands-on than changing the type as they are not directly compatible, I suggest we log a public issue for additional recipe coverage but focus on what was requested for now.

**Review by timtebeek — APPROVED** [review]
Approved after small polish to suppress expected deprecation warnings, and use consistent signature for Refaster recipe templates.

**Pull request merged**